impl/sql/mutationstorage/queue.go

```diff
@@ -125,37 +140,63 @@ func (m *Mutations) send(ctx context.Context, domainID string, logID int64, mDat
 	return tx.Commit()
 }
 
-// HighWatermarks returns the highest timestamp for each log in the mutations table.
-func (m *Mutations) HighWatermarks(ctx context.Context, domainID string) (map[int64]int64, error) {
+// HighWatermarks returns the highest watermark that is less than or equal to batchSize items
+// greater than the watermark for that log in starts, for each log for domainID.
+// The total size will be numLogs * batchSize.
+func (m *Mutations) HighWatermarks(ctx context.Context, domainID string,
+	starts map[int64]int64, batchSize int32) (map[int64]int32, map[int64]int64, error) {
 	watermarks := make(map[int64]int64)
-	rows, err := m.db.QueryContext(ctx,
-		`SELECT LogID, Max(Time) FROM Queue WHERE DomainID = ? GROUP BY LogID;`,
-		domainID)
+	counts := make(map[int64]int32)
+
+	filterForWritable := false
+	logIDs, err := m.logList(ctx, domainID, filterForWritable)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	defer rows.Close()
-	for rows.Next() {
-		var logID, watermark int64
-		if err := rows.Scan(&logID, &watermark); err != nil {
-			return nil, err
+	for _, logID := range logIDs {
+		start := starts[logID]
+		count, high, err := m.highWatermark(ctx, domainID, logID, start, batchSize)
+		if err != nil {
+			return nil, nil, status.Errorf(codes.Internal,
+				"highWatermark(%v/%v, start: %v, batch:%v): %v",
+				domainID, logID, start, batchSize, err)
 		}
-		watermarks[logID] = watermark
+		watermarks[logID] = high
+		counts[logID] = count
 	}
-	if err := rows.Err(); err != nil {
-		return nil, err
-	}
-	return watermarks, nil
+	return counts, watermarks, nil
+}
+
+// highWatarmark returns the highest watermark in logID that is less than or
+// equal to batchSize items greater than start.
+func (m *Mutations) highWatermark(ctx context.Context, domainID string, logID,
```

Comment:
Following our in-person discussion, it would be clearer / more flexible to have only this call behind the interface. The "global" batching strategy can stay outside.